File: src/main/java/com/enthusia/koth/infrastructure/integration/VaultEconomyAdapter.java
package com.enthusia.koth.infrastructure.integration;

import com.enthusia.koth.application.ports.EconomyPort;
import com.enthusia.koth.application.ports.TransactionResult;
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import net.milkbowl.vault.economy.Economy;
import org.bukkit.OfflinePlayer;
import org.bukkit.plugin.RegisteredServiceProvider;
import org.bukkit.plugin.java.JavaPlugin;

public final class VaultEconomyAdapter implements EconomyPort {
    private final JavaPlugin plugin;
    private Economy economy;

    @SuppressFBWarnings(value = "EI_EXPOSE_REP2", justification = "JavaPlugin is the Bukkit-owned service lookup source.")
    public VaultEconomyAdapter(JavaPlugin plugin) {
        this.plugin = plugin;
        reload();
    }

    public void reload() {
        if (plugin.getServer().getPluginManager().getPlugin("Vault") == null) {
            economy = null;
            return;
        }
        RegisteredServiceProvider<Economy> registration = plugin.getServer().getServicesManager().getRegistration(Economy.class);
        economy = registration == null ? null : registration.getProvider();
    }

    @Override
    public boolean isAvailable() {
        return economy != null;
    }

    @Override
    public boolean has(OfflinePlayer player, double amount) {
        return economy != null && economy.has(player, amount);
    }

    @Override
    public TransactionResult withdraw(OfflinePlayer player, double amount, String reason) {
        if (economy == null) return TransactionResult.failure("Vault economy is unavailable.");
        var response = economy.withdrawPlayer(player, amount);
        return response.transactionSuccess() ? TransactionResult.success(reason) : TransactionResult.failure(response.errorMessage);
    }

    @Override
    public TransactionResult deposit(OfflinePlayer player, double amount, String reason) {
        if (economy == null) return TransactionResult.failure("Vault economy is unavailable.");
        var response = economy.depositPlayer(player, amount);
        return response.transactionSuccess() ? TransactionResult.success(reason) : TransactionResult.failure(response.errorMessage);
    }
}

File: src/main/kotlin/net/badgersmc/ek/infrastructure/vault/VaultEconomyAdapter.kt
package net.badgersmc.ek.infrastructure.vault

import net.badgersmc.ek.application.PlayerEconomy
import net.milkbowl.vault.economy.Economy
import org.bukkit.Bukkit
import org.bukkit.plugin.java.JavaPlugin
import java.util.UUID

class VaultEconomyAdapter(private val plugin: JavaPlugin) : PlayerEconomy {
    private fun provider(): Economy? =
        plugin.server.servicesManager.getRegistration(Economy::class.java)?.provider

    override fun isAvailable(): Boolean = provider() != null

    override fun balance(playerId: UUID): Double =
        provider()?.getBalance(Bukkit.getOfflinePlayer(playerId)) ?: 0.0

    override fun withdraw(playerId: UUID, amount: Double): Boolean {
        if (amount <= 0.0) return true
        return provider()?.withdrawPlayer(Bukkit.getOfflinePlayer(playerId), amount)?.transactionSuccess() == true
    }

    override fun deposit(playerId: UUID, amount: Double): Boolean {
        if (amount <= 0.0) return true
        return provider()?.depositPlayer(Bukkit.getOfflinePlayer(playerId), amount)?.transactionSuccess() == true
    }
}
